src/main/java/com/google/jenkins/plugins/credentials/oauth/JsonServiceAccountConfig.java

```diff
@@ -53,36 +62,93 @@
   private static final long serialVersionUID = 6818111194672325387L;
   private static final Logger LOGGER =
       Logger.getLogger(JsonServiceAccountConfig.class.getSimpleName());
-  private String jsonKeyFile;
+  @Nonnull
+  private final String filename;
+  @Nonnull
+  private final SecretBytes secretJsonKey;
+  @Deprecated   // for migration purpose
+  @CheckForNull
+  private transient String jsonKeyFile;
   private transient JsonKey jsonKey;
 
+  /**
+   * @param jsonKeyFile uploaded json key file
+   * @param filename
+   *     previous json key file name.
+   *     used if jsonKeyFile is not provided.
+   * @param secretJsonKey
+   *     previous json key file content.
+   *     used if jsonKeyFile is not provided.
+   * @since 0.7
+   */
   @DataBoundConstructor
   public JsonServiceAccountConfig(FileItem jsonKeyFile,
-      String prevJsonKeyFile) {
+      String filename, SecretBytes secretJsonKey) {
     if (jsonKeyFile != null && jsonKeyFile.getSize() > 0) {
       try {
         JsonKey jsonKey = JsonKey.load(new JacksonFactory(),
                 jsonKeyFile.getInputStream());
-        if (jsonKey.getClientEmail() != null &&
-                jsonKey.getPrivateKey() != null) {
-          try {
-            this.jsonKeyFile = writeJsonKeyToFile(jsonKey);
-          } catch (IOException e) {
-            LOGGER.log(Level.SEVERE, "Failed to write json key to file", e);
-          }
+        if (jsonKey.getClientEmail() == null ||
+                jsonKey.getPrivateKey() == null) {
+          throw new IllegalArgumentException("Invalid json key file");
```

Comment:
I cloned and ran this branch locally. Unfortunately throwing these exceptions results in the Jenkins Devil error page, which is definitely something that shouldn't happen. I think a better user experience would be seeing a verification message that tells the user that the file is invalid. My suggestion is to run these checks both here and in a `doCheckJsonKeyFile` verification method on the descriptor, and use these error strings as the message in the `FormValidation.error`, and go back to logging rather than throwing exceptions for these errors.

There would then be no difference in behavior as we still don't actually save the file. When the user reopens these credentials and presses update to reconfigure them, then that will lead to a similar validation message.

Reply:
I don’t think `doCheckXxxx` is applicable to file uploads as the content of the file is sent to the server only when the entire form is submitted.

Reply:
Ah, I guess that's true. I suppose at the very least you can check whether the filename is null/empty without verifying the file contents. And then I would prefer to have logging and create the bad file here than throw these exceptions.

Reply:
OK. I’ll update the request and make it behave as before, just leaving system logs.